### PR TITLE
Fix handling of sender granularities:

### LINKED ISSUE #82: Sender granularities aren't implemented correctly
use_name_in_from isn't implemented correctly.  parse_pcap*.py still splits up emails by directory, where the directory has name + address.  Then we create one detector per directory.  This means that if use_name_in_from=1 and we have one name John Doe with two email addresses john1@gmail.com and john2@gmail.com, they'll be in two different directories on the filesystem, so they'll be two different detector instances.  Consequently we fail to derive a single sender profile for John Doe -- instead, we generate two sender profiles.  In other words, we handle the emails from John Doe <john1@gmail.com> separately from emails from John Doe <john2@gmail.com>; instead, we should be aggregating all those emails and deriving a single sender profile for that one sender, assuming use_name_in_from=1.

We need to merge all emails from a single sender.  I see two ways to implement this:

1. Modify parse_pcap*.py, to store in a different directory structure where the directory depends only on the name if use_name_in_from=1.

2. Modify generate_features.py, to merge all the emails from all the directories associated with a single sender.  This sounds tedious (e.g., we have to process the emails in chronologically order, so we'll have to sort the emails).

Approach #1 seems easier to implement but either could potentially work.

### LINKED ISSUE #83: Make From: line parsing more robust
We should probably adopt more robust and consistent code for parsing the From: line.

Consistency: In parse_pcap.py, we just use `email_utils.parseaddr()`.  In the detectors, we use `extract_from()` in detector.py, which uses our own bespoke code.

Robustness: There are weird things that can occur in From: lines.  Grant came up with some code that he fine-tuned to account for them; we might want to adopt his code.  It is basically the following:

```
import email.utils as email_utils
import re
import sys

# EMAIL_REGEX = (Optional <)(Something w/o spaces,<,>)@(Something w/o spaces,<,>)(Optional >)
EMAIL_REGEX = "<?" + "(" + "[^\s<>,]+@[^\s<>,]+" + ")" + ">?"

def parse_single_person(person):
    """Parse a string in single person format '*Name* *Email Addr*'."""
    realname, email_addr = email_utils.parseaddr(person)
    realname = realname.strip().lower()  # remove whitespace from ends
    email_addr = email_addr.strip().lower()  # remove whitespace from ends

    if realname == '' and email_addr == '':
        # If Python's email util lib fails to parse, it should return empty
        # strings for both realname and email_addr.
        sys.stderr.write(
            "%s: Email parser couldn't handle this person: %s" %
            (ERROR_PERSON, person))
        return ERROR_PERSON
    elif email_addr == '' or not re.match(EMAIL_REGEX, email_addr):
        # Case of broken email_addr parsing or field with no email addr.
        sys.stderr.write(
            ("%s: Email parser (wrongly) says this is an email addr: *%s*. "
             "Full field = %s." % (ERROR_PERSON, email_addr, person)))
        return ERROR_PERSON

    if realname == '':
        # If the field is nameless, just use the email addr as the human name
        realname = email_addr

    return (realname, email_addr)
```

He also replaces all 0x09 characters with a space before calling `parse_single_person()` on it.  I'm not sure why.

Looking through this, I suspect that we can probably introduce a utility function:

```
def parse_sender(sender):
    sender = sender.replace("\x09", " ")
    realname, email_addr = email_utils.parseaddr(sender)
    email_addr = email_addr.strip().lower()
    realname = realname.strip().lower()

    if realname == '':
        # If the field is nameless, just use the email addr as the human name
        realname = email_addr

    return (realname, email_addr)
```

and then in Detector's extract_from(), if use_name_in_from=1, we can probably use something like:

```
def extract_name(headerval):
    return parse_sender(headerval)[0]
```

and if use_name_in_from=0, we can probably use something like:

```
def extract_full_from(headerval):
    return headerval.replace("\x09", " ").strip().lower()
```

and in parse_pcap.py we can probably use something like:

```
def dir_for_sender(sender):
    name, addr = parse_sender(sender)
    name = name.replace("/", "")
    addr = addr.replace("/", "")
    if name = "":
        name = "noname"
    if addr = "":
        addr = "noaddr"
    first_subdir = name[:3]
    second_subdir = name[3:6]
    if second_subdir = "":
        second_subdir = "none"
    if use_name_in_from:
        return '{}/{}/{}/{}'.format(OUTPUT_DIRECTORY, first_subdir, second_subdir, name)
    else:
        return '{}/{}/{}/{}/{}'.format(OUTPUT_DIRECTORY, first_subdir, second_subdir, name, addr)
```

This would also fix issue #82 as a side effect.

What do you think?

Cc: @mikeaboody , @nexusapoorvacus

### PR DESCRIPTION
**Pull request by davidwagner**
choose directory structure so that no sender's emails are
split across more than one directory.
Implement more robust email address parsing.
Fixes #82, fixes #83.

### CODE REVIEW

**Comment by mikeaboody** [comment]
Just for reference, before having tried this, I get the following validation accuracy and confusion matrix: 

INFO     01/26/2017 22:47:03  Confusion matrix - True positives: 1241.0, False positives: 7.0, False negatives: 2051.0, True negatives: 3285.0
INFO     01/26/2017 22:47:03  Validation Accuracy: 0.687424058323
INFO     01/26/2017 22:47:03  False positive rate: 0.00212636695018
INFO     01/26/2017 22:47:03  False negative rate: 0.623025516403

**Comment by mikeaboody** [comment]
New validation accuracy and confusion matrix:
INFO     01/26/2017 23:02:10  Confusion matrix - True positives: 961.0, False positives: 8.0, False negatives: 2321.0, True negatives: 3274.0
INFO     01/26/2017 22:57:01  Validation Accuracy: 0.644119439366
INFO     01/26/2017 22:57:01  False positive rate: 0.00091407678245
INFO     01/26/2017 22:57:01  False negative rate: 0.710847044485

**Comment by mikeaboody** [comment]
Second run in case noise exists:

INFO     01/26/2017 23:02:10  Confusion matrix - True positives: 961.0, False positives: 8.0, False negatives: 2321.0, True negatives: 3274.0
INFO     01/26/2017 23:02:10  Validation Accuracy: 0.645185862279
INFO     01/26/2017 23:02:10  False positive rate: 0.00243753808653
INFO     01/26/2017 23:02:10  False negative rate: 0.707190737355

**Comment by mikeaboody** [comment]
Not sure if this is statistical noise but I notice the false negative rate is 10% higher than it was before. Any reason this might be the case?

**Comment by mikeaboody** [comment]
By the way, all of the aformentioned runs had use_name_in_from = 1, which might explain it? let me try it with use_name_in_from = 0.

**Comment by mikeaboody** [comment]
With use_name_in_from = 0:

INFO     01/26/2017 23:09:03  Confusion matrix - True positives: 1166.0, False positives: 6.0, False negatives: 2116.0, True negatives: 3276.0
INFO     01/26/2017 23:09:03  Validation Accuracy: 0.676721511274
INFO     01/26/2017 23:09:03  False positive rate: 0.0018281535649
INFO     01/26/2017 23:09:03  False negative rate: 0.644728823888

I guess this makes sense as it's doing what it did before: using the full from. Does any of this seem suspicious to you @davidwagner?

**Comment by davidwagner** [comment]
This looks believable to me.  Thanks for the last run with use_name_in_from. It doesn't seem suspicious to me.  OK to merge?
